### PR TITLE
Add local file upload functionality to adapter.

### PR DESCRIPTION
**Pull request by danthelion**
Partly resolves https://github.com/brooklyn-data/dbt_artifacts/issues/61

### Description

Add local file upload functionality to the adapter. The main motivation for this is to make [dbt_artifacts](https://github.com/brooklyn-data/dbt_artifacts) compatible with Databricks. In order to do so we need to get the generated artifacts first to `dbfs`, from where to dbt package will be able to pick them up as sources for the downstream models. Ideally this upload process should happen in the adapter so the whole process in environment-independent and will do the same on dbt Cloud, local runs, etc.

As I haven't found an official Python Databricks API client, I went with [databricks-api](https://github.com/crflynn/databricks-api), which generates the client based on the internals of the official `databricks-cli` tool.

To use via command line

```shell
dbt -d run-operation databricks__upload_file --args "{local_file_path: './sample_files/manifest.js'}"

dbt -d run-operation databricks__upload_file --args "{local_file_path: './sample_files/manifest.js', overwrite: True}"
```

or as a pre-hook

```shell
{{
    config(
        pre_hook=[
            "{{ databricks__upload_file(
                local_file_path='./sample_files/manifest.json'
            ) }}",
            "{{ databricks__upload_file(
                local_file_path='./sample_files/manifest.json',
                overwrite=True
            ) }}"
        ]
    )
}}
```

### Checklist

- [*] I have run this code in development and it appears to resolve the stated issue
- [*] This PR includes tests, or tests are not required/relevant for this PR
- [ ] I have updated the `CHANGELOG.md` and added information about my change to the "dbt-databricks next" section.

### CODE REVIEW

**Comment by robertsaxby** [comment]
@superdupershant ptal

File: tests/integration/upload_file/test_upload_file.py [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+import hashlib
+import os
+
+import yaml
+
+from tests.integration.base import DBTIntegrationTest, use_profile
+
+
+class TestUploadFile(DBTIntegrationTest):
+    local_file_path = './files/run/manifest.json'
```
Comment (ueshin) [inline tests/integration/upload_file/test_upload_file.py:11]: Does the test file need to be so big? I guess we can use just a small file like 1 line or so.

**Review by ueshin — COMMENTED** [review]
@danthelion Thanks for the contribution!
I left some comments and questions. Could you take a look? Thanks!

File: dbt/include/databricks/macros/adapters.sql [hunk 2]
```diff
@@ -163,3 +163,7 @@
 
     {% do return(tmp_relation) %}
 {% endmacro %}
+
+{% macro databricks__upload_file(local_file_path, dbfs_file_path, overwrite, contents, headers) %}
```
[the remark this answers is not in the document]
Comment (danthelion) [reply dbt/include/databricks/macros/adapters.sql:167]: Yes, here is the current Snowflake-only implementation: https://github.com/brooklyn-data/dbt_artifacts/blob/main/macros/upload_artifacts.sql, as it allows to copy straight from local files it doesn't need to go to the adapter.

File: dbt/adapters/databricks/connections.py [hunk 3]
```diff
@@ -300,3 +296,9 @@ def open(cls, connection: Connection) -> Connection:
         connection.handle = handle
         connection.state = ConnectionState.OPEN
         return connection
+
+    @staticmethod
+    def get_user_agent_entry() -> str:
+        dbt_databricks_version = __version__.version
+        dbt_invocation_env = os.getenv(DBT_INVOCATION_ENV) or "manual"
+        return f"dbt-databricks/{dbt_databricks_version}; {dbt_invocation_env}"
```
[the remark this answers is not in the document]
Comment (danthelion) [reply dbt/adapters/databricks/connections.py]: Removed, sorry for that, accidentally left it in from an initial trial.

File: dbt/adapters/databricks/impl.py [hunk 4]
```diff
@@ -232,3 +231,33 @@ def _catalog(self, catalog: Optional[str]) -> Iterator[None]:
...
+            token=creds.token
+        )
+
+        dbapi_client.dbfs.put(
+            path=dbfs_file_path,
+            src_path=local_file_path,
+            overwrite=overwrite,
+            contents=contents,
+            headers=headers
+        )
```
[the remark this answers is not in the document]
Comment (danthelion) [reply dbt/adapters/databricks/impl.py]: I agree, refactored it so the `DatabricksConnectionManager` is responsible for creating the API connection and sending the query. Although I am not sure how an `_execute_api` would look and what kind of trace logs you mean. Could you elaborate a bit?

File: tests/integration/upload_file/test_upload_file.py [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+        })
+
+        result = self.run_dbt(['run-operation', 'upload_file', '--args', upload_args], expect_pass=expect_pass)
+        return result
+
+    def test_upload_file_no_overwrite(self):
+        dbt_run_result = self.upload_file(overwrite=True)
+        self.assertTrue(dbt_run_result.success)
+        dbt_run_result = self.upload_file(overwrite=False, expect_pass=False)
+        self.assertFalse(dbt_run_result.success)
```
[the remark this answers is not in the document]
Comment (danthelion) [reply tests/integration/upload_file/test_upload_file.py]: Done

File: tests/integration/upload_file/test_upload_file.py (hunk 1 above)
Comment (danthelion) [reply tests/integration/upload_file/test_upload_file.py]: True, replaced it with a small file instead!

**Comment by danthelion** [comment]
Thanks for the comments @ueshin! Added my responses & new commits. The only thing I'm not exactly sure about is the implementation of an `_execute_api` wrapper, I am not yet that familiar with dbt adapter patterns (especially this one) but I would love to learn how to fit this into the existing codebase properly.

**Comment by benc-db** [comment]
@danthelion is this still an outstanding ask, or has it been solved some other way in the meantime?  Apologies for this PR getting lost in the shuffle as we've switched maintainers a few times in the last year.

**Comment by benc-db** [comment]
Closing as stale with no response from Dan.

**Pull request closed without merging**